sample: implemented Buckets() function

mvisonneau opened the pull request:
This is a new feature which should allow us to export buckets on top of precomputed percentiles.

This makes more sense to leverage these values when we are aggregating them amongst multiple sources. It would also
allow us to represent our data through the form of distribution charts, which would enhance our troubleshooting
capabilities.

mvisonneau committed:
sample: implemented Buckets() function

This is a new feature which should allow us to export buckets on top of precomputed percentiles.
This makes more sense to leverage these values when we are aggregating them amongst multiple sources. It would also
allow us to represent our data through the form of distribution charts, which would enhance our troubleshooting
capabilities.

poly2d commented on sample_test.go:
```diff
@@ -301,6 +301,20 @@ func testExpDecaySampleStatistics(t *testing.T, s Sample) {
 	}
+
+	buckets := s.Buckets([]int64{100, 500, 20000})
+	if buckets[0] != 0 {
+		t.Errorf("bucket (100): %v != 0\n", buckets[0])
+	}
+	if buckets[1] != 8 {
+		t.Errorf("bucket (500): %v != 8\n", buckets[1])
+	}
+	if buckets[2] != 92 {
+		t.Errorf("bucket (20000): %v != 92\n", buckets[2])
+	}
+	if buckets[3] != 0 {
+		t.Errorf("bucket (+Inf): %v != 0\n", buckets[3])
+	}
```
These expected values are artifacts of the particular seed used, correct?

mvisonneau replied:
absolutely, I tried to follow the existing approach which was implemented for `Percentiles()`

poly2d commented on sample_test.go:
```diff
@@ -329,6 +343,20 @@ func testUniformSampleStatistics(t *testing.T, s Sample) {
+	}
+	if buckets[3] != 0 {
+		t.Errorf("bucket (+Inf): %v != 0\n", buckets[3])
+	}
```
For test coverage, we could also add a case that tests the case where we have some values higher than the maximum bucket provided.

poly2d commented on sample.go:
```diff
@@ -296,6 +308,29 @@ func SamplePercentiles(values int64Slice, ps []float64) []float64 {
 	return scores
 }
 
+// SampleBuckets returns a slice of arbitrary buckets of the slice of
+// int64.
+func SampleBuckets(values int64Slice, buckets []int64) []int64 {
```
I think it is assumed in the func that the input `buckets` is sorted in ascending order, no? Should we add a comment to clarify that, or a check to enforce it?

mvisonneau replied:
absolutely, let's add that!

poly2d reviewed (approved):
Thanks! Left some questions

mvisonneau commented:
- I added the sort method
- Enhanced the impacted test helper functions
- Refactored the loop incrementing the buckets to make it follow the same logic as the openmetrics/prometheus one: https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations
